app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt

```diff
@@ -38,76 +71,199 @@ class FireDatabase : Database() {
+            //check if the username is available
+            isUsernameAvailable(username).thenAccept { isAvailable ->
+                if (isAvailable) {
+                    //add the link between the username and the userId
+                    val usernameToUserId = HashMap<String, String>()
+                    usernameToUserId.put(username, userId)
+                    accessUsernameToUserIdFile().updateChildren(usernameToUserId as Map<String, Any>)
+                        .addOnSuccessListener {
+                            //add the users account to the database and the username to the user account
+                            val userAccount = HashMap<String, String>()
+                            userAccount.put(usernameFile, username)
+                            database.child(usersProfileFileName).child(userId)
+                                .updateChildren(userAccount as Map<String, Any>)
+                                .addOnSuccessListener { future.complete(true) }
+                                .addOnFailureListener {
+                                    future.completeExceptionally(Exception("Impossible to create the user account."))
+                                }
+                        }
+                        .addOnFailureListener {
+                            future.completeExceptionally(java.lang.Error("Impossible to find the link between the username and the userId to the Database."))
+                        }
+                } else {
+                    future.completeExceptionally(java.lang.Error("The username is not available !"))
+                }
+            }
+        }
+        return future
+    }
+
+    override fun initUserProfile(userModel: UserModel): CompletableFuture<Boolean> {
+        val userData = HashMap<String, Any>()
+        userData.put(emailFile, userModel.getEmailAddress())
+        userData.put(firstnameFile, userModel.getFirstname())
+        userData.put(surnameFile, userModel.getSurname())
+        userData.put(dateOfBirthFile, userModel.getDateOfBirth().toEpochDay())
+        userData.put(distanceGoalFile, userModel.getDistanceGoal())
+        userData.put(activityTimeGoalFile, userModel.getActivityTime())
+        userData.put(nbOfPathsGoalFile, userModel.getNumberOfPathsGoal())
+
+        return updateUserData(userData)
+    }
+
+    override fun getUserAccount(userId: String): CompletableFuture<UserModel> {
+        val future = CompletableFuture<UserModel>()
+
+        database.child(usersProfileFileName).child(userId).get().addOnSuccessListener { userData ->
+            if(userData == null) {
+                future.completeExceptionally(java.lang.Error("There is no user account corresponding to this userId."))
+            }else{
+                if(USER_AUTH==null){
+                    future.completeExceptionally(java.lang.Error("There is no user logged on the app."))
+                }else{
+                    val username = userData.child(usernameFile).value
```

Comment:
It makes the function quite complex. Might be worth it to refactor in a `userFileToUserModel` function maybe?

Reply:
I don't think it will be useful since this code is only present in this function. So create an helper wouldn't have a great interest.

Reply:
It would make the function a lot more readable.

Reply:
It would give the function below
```kotlin
override fun getUserAccount(userId: String): CompletableFuture<UserModel> {
    val future = CompletableFuture<UserModel>()

    accessUserAccountFile(userId).get().addOnSuccessListener { userData ->
        val userData = fileToUserModel(userData)
        if (userData != null) {
            future.complete(userData)
        }
        else {
            future.completeExceptionally(it)
        }
    }.addOnFailureListener{
        future.completeExceptionally(it)
    }
    return future
}
```

Reply:
It's done thanks